File: memory-domain/src/main/java/com/memory/domain/member/Member.java
package com.memory.domain.member;

import com.memory.domain.BaseTimeEntity;
import com.memory.domain.file.File;
import com.memory.domain.map.Map;
import com.memory.domain.memberlink.MemberLink;
import com.memory.domain.relationship.Relationship;
import jakarta.persistence.*;
import lombok.*;

import java.util.ArrayList;
import java.util.List;

@ToString
@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private String nickname;

    private String email;

    private String password;

    @OneToOne(mappedBy = "member")
    private File file;

    @OneToOne(mappedBy = "member")
    private Relationship relationship;

    @OneToOne(mappedBy = "relatedMember")
    private Relationship relatedRelationship;    // 내가 받은 관계

    @Enumerated(EnumType.STRING)
    private MemberType memberType;

    @OneToMany(mappedBy = "member", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Map> maps = new ArrayList<>();

    @OneToMany(mappedBy = "member", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MemberLink> memberLinks = new ArrayList<>();

    public Member(String name, String nickname, String email, String password) {
        this.name = name;
        this.nickname = nickname;
        this.email = email;
        this.password = password;
        this.memberType = MemberType.MEMBER;
    }

    public Member(String name, String nickname, String email, String password, MemberType memberType) {
        this.name = name;
        this.nickname = nickname;
        this.email = email;
        this.password = password;
        this.memberType = memberType;
    }

    public void update(String nickname, File file) {
        this.nickname = nickname;
        this.updateFile(file);
    }

    // 연관관계 편의 메서드
    private void updateFile(File file) {
        if (this.file != null) {
            this.file.updateMember(null);
        }
        this.file = file;
        file.updateMember(this);
    }

    public void updatePassword(String password) {
        if (password != null) {
            this.password = password;
        }
    }

    // MemberLink 연관관계 편의 메서드
    public void addMemberLink(MemberLink memberLink) {
        this.memberLinks.add(memberLink);
    }
}

File: memory-domain/src/main/kotlin/com/memory/domain/member/Member.kt
package com.memory.domain.member

import com.memory.domain.BaseTimeEntity
import com.memory.domain.file.File
import com.memory.domain.map.Map
import com.memory.domain.memberlink.MemberLink
import com.memory.domain.relationship.Relationship
import jakarta.persistence.*

@Entity
@Table(name = "member")
class Member(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    var name: String,

    var nickname: String,

    var email: String,

    var password: String,

    @Enumerated(EnumType.STRING)
    var memberType: MemberType = MemberType.MEMBER,

    @OneToOne(mappedBy = "member")
    var file: File? = null,

    @OneToOne(mappedBy = "member")
    var relationship: Relationship? = null,

    @OneToOne(mappedBy = "relatedMember")
    var relatedRelationship: Relationship? = null,

    @OneToMany(mappedBy = "member", cascade = [CascadeType.ALL], orphanRemoval = true)
    var maps: MutableList<Map> = mutableListOf(),

    @OneToMany(mappedBy = "member", cascade = [CascadeType.ALL], orphanRemoval = true)
    var memberLinks: MutableList<MemberLink> = mutableListOf()
) : BaseTimeEntity() {

    constructor(name: String, nickname: String, email: String, password: String) : this(
        id = null,
        name = name,
        nickname = nickname,
        email = email,
        password = password
    )

    fun update(nickname: String, file: File?) {
        this.nickname = nickname
        file?.let { updateFile(it) }
    }

    private fun updateFile(file: File) {
        this.file?.updateMember(null)
        this.file = file
        file.updateMember(this)
    }

    fun updatePassword(password: String?) {
        password?.let { this.password = it }
    }

    fun addMemberLink(memberLink: MemberLink) {
        this.memberLinks.add(memberLink)
    }
}
